### DISCUSSION TITLE
How to define a 1-1 and 1-M symmetrical realtion?

### DISCUSSION THREAD

**Discussion by wsw70** [ent/ent#2391] [Q&A]
I  am an amateur developer and just started with ORM in Go.  Ent seems to be great but I have some concerns about my understanding of relationships.

I have a schema `Entry` (think ticket or case) that is supposed to have only one `State` (open, closed, pending, ...). A specific `State will exist on many `Entry`.

The natural way of thinking for me is to use `Entry` as a "master data" container, and `State` would be one of its attributes. But I cannot find any relationships that would match this in https://entgo.io/docs/schema-edges#relationship

I believe that what I am looking for is the graph below (the second graph should be rotated 180° (the arrows should go to the left to match the graphs in the documentation) but I do not know how to do that in mermaid)

```mermaid
flowchart LR
Entry-->|1|state
state-->|1|State
```
```mermaid
flowchart LR
State-->|1| entries
entries-->|M|Entry
```

Is this correct? And if so - how can I express that in the code?

**Comment by idc77** [comment]
Personally I would use an enum for state.
```
type Entry {
  Type Enum{"ticket","case"}
  State Enum{"open","closed","pending"}
}
```

aka

```go
func (Entry) Fields() []ent.Field {
	return []ent.Field{
		field.UUID("id", uuid.UUID{}).Annotations(
			entsql.Annotation{Default: "gen_random_uuid()"},
		).Unique(),
		field.Time("create_time").Default(time.Now).Immutable(),
		field.Time("update_time").Default(time.Now).UpdateDefault(time.Now).Optional(),
		field.Enum("type").Values("ticket","case"),
		field.Enum("state").Values("open","closed","pending").Default("open"),
	}
}
```

As for your question, you can't have 1:1 and 1:m inverse.
m:1 or 1:m
or
1:1 and 1:1
or
m:m

If you just need a separate entity for State,
In Entry, State is a m:1 (from many state select one)
In State, Entry is a 1:m (my one state has many entry)
```go

func (Entry) Fields() []ent.Field {
	return []ent.Field{
		field.UUID("id", uuid.UUID{}).Annotations(
			entsql.Annotation{Default: "gen_random_uuid()"},
		).Unique(),
		field.Time("create_time").Default(time.Now).Immutable(),
		field.Time("update_time").Default(time.Now).UpdateDefault(time.Now).Optional(),
		field.UUID("state_id", uuid.UUID{}).Optional(),
	}
}

func (Entry) Edges() []ent.Edge {
	return []ent.Edge{
		edge.From("state", State.Type).Ref("entries").Field("state_id").Unique(),
	}
}
```

```go
func (State) Edges() []ent.Edge {
	return []ent.Edge{
		edge.To("entries", Entry.Type)
	}
}
```

But you should really use an enum imho.
